precompiles/parachain-staking/src/lib.rs

```diff
@@ -60,167 +78,43 @@ where
-
-		log::trace!(target: "staking-precompile", "Gonna call with origin {:?}", origin);
-
-		match outer_call.dispatch(Some(origin).into()) {
-			Ok(post_info) => {
-				let gas_used = Runtime::GasWeightMapping::weight_to_gas(
-					post_info.actual_weight.unwrap_or(info.weight),
-				);
-				Ok(PrecompileOutput {
-					exit_status: ExitSucceed::Stopped,
-					cost: gas_used,
-					output: Default::default(),
-					logs: Default::default(),
-				})
-			}
-			Err(e) => {
-				let error_message = format!(
-					"Parachain staking call via EVM failed with dispatch error: {:?}",
-					e
-				);
-				Err(error(error_message))
-			}
-		}
-	}
-}
-
-/// Parses an H160 account address from a 256 bit (32 byte) buffer. Only the last 20 bytes are used.
-fn parse_account(input: &[u8]) -> Result<H160, ExitError> {
-	const PADDING_SIZE_BYTES: usize = 12;
-	const ACCOUNT_SIZE_BYTES: usize = 20;
-	const TOTAL_SIZE_BYTES: usize = PADDING_SIZE_BYTES + ACCOUNT_SIZE_BYTES;
-
-	if input.len() != TOTAL_SIZE_BYTES {
-		log::trace!(target: "staking-precompile",
-			"Unable to parse address. Got {} bytes, expected {}",
-			input.len(),
-			TOTAL_SIZE_BYTES,
-		);
-		return Err(error("Incorrect input length for account parsing"));
-	}
-
-	Ok(H160::from_slice(
-		&input[PADDING_SIZE_BYTES..TOTAL_SIZE_BYTES],
-	))
-}
-
-/// Parses an amount of ether from a 256 bit (32 byte) slice. The balance type is generic.
-fn parse_amount<Balance: TryFrom<U256>>(input: &[u8]) -> Result<Balance, ExitError> {
-	Ok(parse_uint256(input)?
-		.try_into()
-		.map_err(|_| error("Amount is too large for provided balance type"))?)
-}
-
-/// Parses a uint256 value
-fn parse_uint256(input: &[u8]) -> Result<U256, ExitError> {
-	// In solidity all values are encoded to this width
-	const SIZE_BYTES: usize = 32;
-
-	if input.len() != SIZE_BYTES {
-		log::trace!(target: "staking-precompile",
-			"Unable to parse uint256. Got {} bytes, expected {}",
-			input.len(),
-			SIZE_BYTES,
-		);
-		return Err(error("Incorrect input length for uint256 parsing"));
-	}
-
-	Ok(U256::from_big_endian(&input[0..SIZE_BYTES]))
-}
-
-/// Parses Weight Hint: u32 from a 256 bit (32 byte) slice.
-fn parse_weight_hint(input: &[u8]) -> Result<u32, ExitError> {
-	const WEIGHT_HINT_SIZE_BYTES: usize = 32;
-
-	if input.len() != WEIGHT_HINT_SIZE_BYTES {
-		log::trace!(target: "staking-precompile",
-			"Unable to parse weight hint. Got {} bytes, expected {}",
-			input.len(),
-			WEIGHT_HINT_SIZE_BYTES,
-		);
-		return Err(error("Incorrect input length for weight hint parsing"));
+		// Initialize gasometer
+		let mut gasometer = Gasometer::new(target_gas);
+		// Dispatch call (if enough gas).
+		let used_gas =
+			RuntimeHelper::<Runtime>::try_dispatch(origin, call, gasometer.remaining_gas()?)?;
+		gasometer.record_cost(used_gas)?;
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: vec![],
+			logs: vec![],
+		})
```

Comment:
@JoshOrndorff all dispatchables share the dispatch code now, as per this review comment: https://github.com/PureStake/moonbeam/pull/632#discussion_r677457153